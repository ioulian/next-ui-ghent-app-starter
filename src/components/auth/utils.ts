"use server";

import "server-only";

import { signInCredentials } from "@/auth";

export type CredentialsType = {
  username: string;
  password: string;
};

export const login = async (data?: CredentialsType): Promise<void> => {
  if (data) {
    const { username, password } = data;
    await signInCredentials(username, password);
  }
};
