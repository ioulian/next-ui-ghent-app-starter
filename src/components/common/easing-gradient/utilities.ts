import { easingCoordinates } from "easing-coordinates";
import { mix, parse, ColorSpace, sRGB, Lab, serialize } from "colorjs.io/fn";

// If you want to use more color spaces, you'll need to register them first
ColorSpace.register(sRGB);
ColorSpace.register(Lab);

/**
 * Will generate a smooth gradient to be used in css
 *
 * Some info: https://non-boring-gradients.netlify.app/
 * "Ported" from https://github.com/larsenwork/postcss-easing-gradients
 *
 * @param from Start color
 * @param to End color
 * @param easing Easing function
 * @param stops Number of stops to use. Do not go overboard with it!
 * @returns Comma separated color stops to be used in "linear-gradient"
 */
export const createEasingGradient = (
  from: string,
  to: string,
  easing: string = "cubic-bezier(0.5, 0, 0.5, 1)",
  stops: number = 16,
) => {
  const coordinates = easingCoordinates(easing, stops);

  // TS ignore is there because the types for serialize are not correct. This works.
  return coordinates
    .map(
      ({ x, y }) =>
        `${serialize(
          mix(parse(from), parse(to), y, {
            space: "lab",
            outputSpace: "srgb",
          }),
        )} ${Math.abs(x * 100).toFixed(2)}%`,
    )
    .join(", ");
};
