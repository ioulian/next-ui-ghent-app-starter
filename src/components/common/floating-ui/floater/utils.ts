export const roundByDPR = (value: number): number => {
  const dpr = window.devicePixelRatio || 1;
  return Math.round(value * dpr) / dpr;
};
