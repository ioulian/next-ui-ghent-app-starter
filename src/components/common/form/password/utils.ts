import { TranslationKeys, zxcvbn, zxcvbnOptions } from "@zxcvbn-ts/core";

const loadOptions = async () => {
  const zxcvbnCommonPackage = await import(
    /* webpackChunkName: "zxcvbnCommonPackage" */ "@zxcvbn-ts/language-common"
  );

  return {
    dictionary: {
      ...zxcvbnCommonPackage.dictionary,
    },
    graphs: zxcvbnCommonPackage.adjacencyGraphs,
  };
};

export const validatePassword = async (
  value: string,
  timeEstimationTranslations: TranslationKeys["timeEstimation"],
) => {
  const options = await loadOptions();
  zxcvbnOptions.setOptions({
    ...options,
    translations: {
      warnings: {
        straightRow: "",
        keyPattern: "",
        simpleRepeat: "",
        extendedRepeat: "",
        sequences: "",
        recentYears: "",
        dates: "",
        topTen: "",
        topHundred: "",
        common: "",
        similarToCommon: "",
        wordByItself: "",
        namesByThemselves: "",
        commonNames: "",
        userInputs: "",
        pwned: "",
      },
      suggestions: {
        l33t: "",
        reverseWords: "",
        allUppercase: "",
        capitalization: "",
        dates: "",
        recentYears: "",
        associatedYears: "",
        sequences: "",
        repeated: "",
        longerKeyboardPattern: "",
        anotherWord: "",
        useWords: "",
        noNeed: "",
        pwned: "",
      },
      timeEstimation: timeEstimationTranslations,
    },
  });

  return zxcvbn(value);
};
