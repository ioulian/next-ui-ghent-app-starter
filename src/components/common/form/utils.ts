// TODO: maybe we don't need it? useId
export const getDescriptionId = (id: string): string => {
  return `${id}Description`;
};

// TODO: maybe we don't need it? useId
export const getErrorId = (id: string): string => {
  return `${id}Error`;
};

/**
 * Will get correct `aria-describedby` attribute for a form field
 *
 * @param id Id of the form field
 * @param hasDescription Does this formfield have a description
 * @param hasError Does this formfield have an error
 */
export const getAriaDescribedBy = (
  id: string,
  hasDescription: boolean = false,
  hasError: boolean = false,
): string | undefined => {
  const errorId = getErrorId(id);
  const descrId = getDescriptionId(id);

  return hasError
    ? hasDescription
      ? `${errorId} ${descrId}`
      : errorId
    : hasDescription
      ? descrId
      : undefined;
};
