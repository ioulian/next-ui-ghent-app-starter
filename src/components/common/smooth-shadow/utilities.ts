// "Ported" from: https://smoothshadows.com/
import { easingCoordinates } from "easing-coordinates";
import Color from "colorjs.io";
import { parse, ColorSpace, sRGB } from "colorjs.io/fn";

ColorSpace.register(sRGB);

export interface SmoothShadowOptionsType {
  startAlpha: number;
  endAlpha: number;
  startBlur: number;
  endBlur: number;
  layers: number;
  x: number;
  y: number;
  color: string;
  alphaEasing: string;
  blurEasing: string;
  distanceEasing: string;
}

/**
 * Create a smooth shadow composed of different "normal" shadows in css.
 *
 * @returns String to be used in css `box-shadow`
 */
export const createSmoothShadow = ({
  /**
   * Start opacity of the shadow
   */
  startAlpha,

  /**
   * End opacity of the shadow
   */
  endAlpha,

  /**
   * Start blur  of the shadow
   */
  startBlur,

  /**
   * End blur  of the shadow
   */
  endBlur,

  /**
   * Number of "normal" css shadows to generate. More is more smooth, but will tank performance quite fast
   */
  layers,
  x,
  y,
  color,
  alphaEasing,
  blurEasing,
  distanceEasing,
}: SmoothShadowOptionsType): string => {
  const distanceCoordinates = easingCoordinates(distanceEasing, layers);
  const blurCoordinates = easingCoordinates(blurEasing, layers);
  const alphaCoordinates = easingCoordinates(alphaEasing, layers);
  const parsedColor = new Color(parse(color));
  const rgb: [number, number, number] = [
    Math.floor(parsedColor.srgb.r * 255),
    Math.floor(parsedColor.srgb.g * 255),
    Math.floor(parsedColor.srgb.b * 255),
  ];

  return Array(layers)
    .fill(0)
    .map((_, i) => ({
      x: distanceCoordinates[i].y * x,
      y: distanceCoordinates[i].y * y,
      blur: startBlur + blurCoordinates[i].y * (endBlur - startBlur),
      alpha: startAlpha + alphaCoordinates[i].y * (endAlpha - startAlpha),
    }))
    .map(
      ({ x, y, blur, alpha }) =>
        `${x.toFixed(1)}px ${y.toFixed(1)}px ${blur.toFixed(
          2,
        )}px rgba(${rgb.join(", ")}, ${alpha.toFixed(2)})`,
    )
    .join(", ");
};
