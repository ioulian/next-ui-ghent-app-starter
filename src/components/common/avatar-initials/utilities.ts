import { escapeSVG } from "@/utils/svg";

/**
 * Will create a simple image with some text in it. Useful when you don't have any image for an user
 *
 * @param initials String to show in the image
 * @param backgroundColor
 * @param textColor
 * @returns svg string
 */
export const getAvatarInitials = (
  initials: string,
  backgroundColor: string = "currentColor",
  textColor: string = "#fff",
): string => {
  return escapeSVG(
    `<svg width="512" height="512" xmlns="http://www.w3.org/2000/svg"><rect height="512" width="512" y="-1" x="-1" fill="${backgroundColor}"/><text transform="matrix(1 0 0 1 0 0)" font-weight="bold" stroke="#000" xml:space="preserve" font-family="'Inter', sans-serif" font-size="180" stroke-width="0" x="50%" y="54%" dominant-baseline="middle" text-anchor="middle" fill="${textColor}">${initials.toLocaleUpperCase()}</text></svg>`,
  );
};
