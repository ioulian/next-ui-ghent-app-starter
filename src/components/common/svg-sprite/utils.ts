import fs from "fs";

import { getBuildId } from "@/utils/next";

export const getSpritesheetUrl = () =>
  // @ts-expect-error FIXME:
  `${__webpack_public_path__}static/media/sprite-${getBuildId()}.svg`;

export const hasSpriteSheet = () => {
  const fileUrl = `${process.cwd()}/.next/static/media/sprite-${getBuildId()}.svg`;

  if (!fs.existsSync(fileUrl)) {
    return false;
  }

  const file = fs.readFileSync(fileUrl, "utf8");

  return file && typeof file === "string" && file.trim() !== "";
};
