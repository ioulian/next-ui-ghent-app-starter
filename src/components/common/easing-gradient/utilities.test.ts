/* eslint-disable i18next/no-literal-string */

import { easings } from "@/utils/easings";

import { createEasingGradient } from "./utilities";

describe("utilities", () => {
  it("generates a correct easing gradient", () => {
    const value = createEasingGradient(
      "rgba(129,254,233,1)",
      "rgba(246,110,255,1)",
      easings.easeInOutQuad,
      16,
    );
    expect(value).toMatchSnapshot();
  });
});
