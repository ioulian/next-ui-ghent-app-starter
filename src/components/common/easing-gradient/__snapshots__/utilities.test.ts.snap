// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities generates a correct easing gradient 1`] = `"rgb(50.588% 99.608% 91.373%) 0.00%, rgb(51.631% 99.085% 91.488%) 8.04%, rgb(54.415% 97.603% 91.813%) 15.37%, rgb(58.344% 95.283% 92.309%) 22.08%, rgb(62.871% 92.238% 92.939%) 28.28%, rgb(67.588% 88.578% 93.664%) 34.07%, rgb(72.223% 84.406% 94.45%) 39.55%, rgb(76.605% 79.826% 95.263%) 44.83%, rgb(80.629% 74.94% 96.075%) 50.00%, rgb(84.237% 69.853% 96.86%) 55.17%, rgb(87.399% 64.682% 97.596%) 60.45%, rgb(90.1% 59.563% 98.263%) 65.93%, rgb(92.335% 54.669% 98.845%) 71.72%, rgb(94.102% 50.232% 99.325%) 77.92%, rgb(95.393% 46.563% 99.688%) 84.63%, rgb(96.193% 44.057% 99.919%) 91.96%, rgb(96.471% 43.137% 100%) 100.00%"`;
